### PR TITLE
Allowing apps in system namespace

### PR DESCRIPTION
**Pull request by gsps1**
JIRA : https://issues.cask.co/browse/CDAP-13148

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractNamespaceResourceDeleter.java [hunk 1]
```diff
@@ -82,7 +82,10 @@
   @Override
   public void deleteResources(NamespaceMeta namespaceMeta) throws Exception {
     final NamespaceId namespaceId = namespaceMeta.getNamespaceId();
-
+    if (namespaceMeta.getNamespaceId().equals(NamespaceId.SYSTEM)) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractNamespaceResourceDeleter.java]: We don't allow deletion of system namespace, right? Meaning no code path should even lead to this class?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ArtifactHttpHandler.java [hunk 2]
```diff
@@ -682,7 +682,10 @@ public BodyConsumer addArtifact(HttpRequest request, HttpResponder responder,
     throws NamespaceNotFoundException, BadRequestException {
 
     final NamespaceId namespace = validateAndGetNamespace(namespaceId);
-
+    if (namespace.equals(NamespaceId.SYSTEM)) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ArtifactHttpHandler.java]: It's better to do this check in the `AuthorizationArtifactRepository`

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/system/CDAPLogAppender.java [hunk 3]
```diff
@@ -236,7 +236,8 @@ LogPathIdentifier getLoggingPath(Map<String, String> propertyMap) throws Illegal
 
     String namespaceId = propertyMap.get(Constants.Logging.TAG_NAMESPACE_ID);
 
-    if (NamespaceId.SYSTEM.getNamespace().equals(namespaceId)) {
+    if (NamespaceId.SYSTEM.getNamespace().equals(namespaceId)
+      && !propertyMap.containsKey(Constants.Logging.TAG_APPLICATION_ID)) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/system/CDAPLogAppender.java]: What is this for?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java [hunk 4]
```diff
@@ -191,6 +191,12 @@ public DynamicDatasetCache(SystemDatasetInstantiator instantiator,
   public final <T extends Dataset> T getDataset(String namespace, String name, Map<String, String> arguments,
                                                 boolean bypass, AccessType accessType)
     throws DatasetInstantiationException {
+    if (namespace.equals(NamespaceId.SYSTEM.getNamespace()) &&
+      !this.namespace.getNamespace().equals(NamespaceId.SYSTEM.getNamespace())) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java]: shouldn't need to call `getNamespace()` to compare. Should just be `NamespaceId.SYSTEM.equals(this.namespace)`

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/system/CDAPLogAppender.java (hunk 3 above)
Comment (gsps1) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/system/CDAPLogAppender.java]: log collection, if the logs are from system namespace, there is preconditions check to make sure it has "service-id" whereas if these logs are from application in system namespace, they wouldn't have that.

Comment (chtyim) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/system/CDAPLogAppender.java]: Would be good to put a comment there

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/scheduler/SchedulerQueueResolver.java [hunk 5]
```diff
@@ -61,6 +62,9 @@ public String getDefaultQueue() {
    */
   @Nullable
   public String getQueue(Id.Namespace namespaceId) throws IOException, NamespaceNotFoundException {
+    if (namespaceId.equals(NamespaceId.SYSTEM.toId())) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/scheduler/SchedulerQueueResolver.java]: For system namespace, I think it should always use the cdap scheduler queue.

https://github.com/caskdata/cdap/blob/develop/cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java#L909

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 6]
```diff
@@ -169,7 +169,9 @@ public BodyConsumer deploy(HttpRequest request, HttpResponder responder,
     throws BadRequestException, NamespaceNotFoundException {
 
     NamespaceId namespace = validateNamespace(namespaceId);
-
+    if (namespace.equals(NamespaceId.SYSTEM)) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: I think it's better to the check in the `AuthorizationArtifactRepository` to have a central place to enforce the checking.

**Comment by gsps1** [comment]
@chtyim could you please take a look at the updated changes ? thanks!

**Comment by chtyim** [comment]
LGTM. Please rebase to a commit and merge.

**Pull request merged**